### PR TITLE
[Merged by Bors] - Simplify design for `*Label`s

### LINKED ISSUE #4954: Simplify design for `*Label` types
Bevy currently has several different kinds of label type, all of which are essentially duplications of each other. For the sake of b~~r~~evity, I will collectively refer to these as just `Label`.

## What problem does this solve or what need does it fill?

The current design for `Label` is overly complex and "heavy". The design is centered around trait objects and boxing, which makes for poor ergonomics, and adds overhead to simple comparisons between labels. Since this is such a common operation, we should strive to make it as cheap as possible.

## What solution would you like?

To guide the design, let's check out the current docs for `Label`:

> Defines a set of strongly-typed labels for a class of objects

1. It's a label, so it needs text.
2. It's a set, so many different types must be coercible to it.
3. It's strongly typed, so two distinct labels with the same text must not collide.

`Label` is now a data struct, copy-able. It keeps track of the type it was made with and stores a string reference.

```rust
use std::any::TypeId;

#[derive(Clone, Copy, PartialEq, Eq, Hash)]
pub struct Label(TypeId, &'static str);

impl Debug/Display for Label { /* ... */ }

pub trait IntoLabel: Copy + 'static {
    fn into_label(self) -> Label {
        let id = TypeId::of::<Self>();
        let text = self.as_str();
        Self(id, text)
    }
    fn as_str(&self) -> &'static str;
}
```

The `'static` requirement for the text may be alarming to some, but I believe the tradeoff is well worth it. It greatly simplifies the implementation and leads to less code being generated and executed, since there's no generic magic, dynamic dispatch, or even branches. And the restriction shouldn't be a problem in most cases, since Labels are usually unit structs, c-style enums, and `fn` items -- all of these types can easily be represented by a string literal. 

```rust
#[derive(Clone, Copy)]
pub enum SomeLabels {
    Foo,
    Bar
}

// result of #[derive(IntoLabel)]
impl IntoLabel for SomeLabel {
    fn as_str(&self) -> &'static str {
        match self {
            Self::Foo => "Foo",
            Self::Bar => "Bar",
        }
    }
}

fn some_system() {
    // ...
}

// as part of some giant macro to implement for all Fn types...
impl<F: FnMut() + Copy + 'static> IntoLabel for F {
    fn as_str(&self) -> &'static str {
        std::any::type_name::<F>()
    }
}
```

In cases when a user *really* needs runtime `Label`s, they could always just leak a string to get a `'static` reference.

## What alternative(s) have you considered?

Don't change the design. The old design may not be perfect but it has worked so far.

### PR DESCRIPTION
**Pull request by JoJoJet**
# Objective

- Closes #4954 
- Reduce the complexity of the `{System, App, *}Label` APIs.

## Solution

For the sake of brevity I will only refer to `SystemLabel`, but everything applies to all of the other label types as well.

- Add `SystemLabelId`, a lightweight, `copy` struct.
- Convert custom types into `SystemLabelId` using the trait `SystemLabel`.

## Changelog

- String literals implement `SystemLabel` for now, but this should be changed with #4409 .

## Migration Guide

- Any previous use of `Box<dyn SystemLabel>` should be replaced with `SystemLabelId`.
- `AsSystemLabel` trait has been modified.
    - No more output generics.
    - Method `as_system_label` now returns `SystemLabelId`, removing an unnecessary level of indirection.
- If you *need* a label that is determined at runtime, you can use `Box::leak`. Not recommended.

## Questions for later

* Should we generate a `Debug` impl along with `#[derive(*Label)]`?
* Should we rename `as_str()`?
* Should we remove the extra derives (such as `Hash`) from builtin `*Label` types?
* Should we automatically derive types like `Clone, Copy, PartialEq, Eq`?
* More-ergonomic comparisons between `Label` and `LabelId`.
* Move `Dyn{Eq, Hash,Clone}` somewhere else.
* Some API to make interning dynamic labels easier.
* Optimize string representation
    * Empty string for unit structs -- no debug info but faster comparisons
    * Don't show enum types -- same tradeoffs as asbove.

### CODE REVIEW

File: crates/bevy_ecs/src/schedule/state.rs
Comment (JoJoJet) [reply crates/bevy_ecs/src/schedule/state.rs:77]: It looks like this type isn't used anywhere outside of the module it's defined in, so I'm just going to yeet it.

**Comment by alice-i-cecile** [comment]
> I imagine runtime labels will be useful for things like scripting or editor support, and having to leak memory is bad.

Hmm. I suspect we will want this for editor support, but I'm not actually convinced that there's a higher memory cost to leaking the memory versus just allocating a static string at compile time.

> Somewhat related to the point above. I find it important to be able to customize Debug impls of labels. With this change, the Debug string of labels are determined by the as_str() function, which can be a surprising place for users.

Interesting, that's a bit surprising to me. Can you explain a bit more @infmagic2047?

> To be honest I am not convinced that we need such a change. Label comparisons mostly happens when the system schedule changes, which should be very infrequent. Do you have benchmarks showing the performance difference?

I would also like to see benchmarks. This will primarily show up in complex schedule intialization. I think that adding such a benchmark is good to do regardless; I'd split this out into a seperate PR.

It also might show up in compile times, so I'd check that.

> I suggest keeping the old trait name *Label, and use another name (like *LabelId) for the struct, to reduce disruption for users.

100% agree. The existing names are clearer, and there's no need to increase churn.

File: crates/bevy_app/src/lib.rs
Comment (alice-i-cecile) [inline crates/bevy_app/src/lib.rs:32]: As part of these changes, we should get rid of all of the extra derives on these types.

My first impression is that we should have a blanket Debug impl, remove Hash, and keep PartialEq, Eq and Clone.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Definitely revert the `SystemLabel` -> `IntoSystemLabel` change; it's making this PR very tricky to review due to all of the noise. Initial impressions are positive though; I don't think there's any serious loss of functionality.

File: crates/bevy_app/src/lib.rs
Comment (JoJoJet) [reply crates/bevy_app/src/lib.rs:32]: I didn't want to break any user code unnecessarily, and I don't see how the extra derives are hurting anyone. What if we only remove them from private definitions?

**Comment by JoJoJet** [comment]
I'm gonna take this point by point
> The use of &'static str looks worrying to me. I imagine runtime labels will be useful for things like scripting or editor support, and having to leak memory is bad. If the IntoLabel traits are derived in almost all cases, can we use a simple integer in the second field of label here (and generate it automatically with the macro)?

This was actually the original design I had for this before publishing the issue. It's elegant, but the problem is that then the label becomes completely opaque, and you can't even print it at all. You can store an integer *and* the string, but that's pointless since you can just use the `string+TypeID` for equality.

We can always do string interning for editor support, but I think Alice already convinced you.

> Somewhat related to the point above. I find it important to be able to customize Debug impls of labels. With this change, the Debug string of labels are determined by the as_str() function, which can be a surprising place for users.

In practice I don't think there's much use for arbitrary `Debug` impls, in most cases you just want constant text. I can rename the `as_str()` fn to make it more clear though.

> I suggest keeping the old trait name *Label, and use another name (like *LabelId) for the struct, to reduce disruption for users.

Makes sense. Done.

> I would also like to see benchmarks. This will primarily show up in complex schedule intialization. I think that adding such a benchmark is good to do regardless; I'd split this out into a seperate PR. 
> It also might show up in compile times, so I'd check that.

I agree that benchmarks are valuable but I don't really know how to benchmark bevy. Is someone willing to help me out on discord?

> Well, in my case it's mostly a desire to have distinguishable StageLabel names in diagnostics and similar places. In my code there are stages GameStage::Update and UiStage::Update along with the builtin CoreStage::Update, and they all show up as Update with the default Debug impl.

Good idea, on it.

**Comment by JoJoJet** [comment]
Also, I think the `Dyn{Eq, Hash, Clone}` types need a new home. Is that within the scope of this PR?

**Comment by alice-i-cecile** [comment]
> Also, I think the Dyn{Eq, Hash, Clone} types need a new home. Is that within the scope of this PR?

Seems reasonable, but make a new PR :)

**Comment by alice-i-cecile** [comment]
> I agree that benchmarks are valuable but I don't really know how to benchmark bevy. Is someone willing to help me out on discord?

100%. Come say hi in #engine-dev and people will be able to help you out :)

File: crates/bevy_ecs/src/schedule/stage.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/schedule/stage.rs:1630]: This is much clearer!

File: crates/bevy_ecs/src/schedule/state.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/schedule/state.rs:111]: I'm not immediately sure why this is being removed. And can we remove this method entirely? It looks like you've deleted a bunch of uses of it.

File: crates/bevy_macro_utils/src/lib.rs
Comment (alice-i-cecile) [inline crates/bevy_macro_utils/src/lib.rs:131]:
```suggestion
            _ => panic!("Labels cannot contain data."),
```

File: crates/bevy_macro_utils/src/lib.rs
Comment (alice-i-cecile) [inline crates/bevy_macro_utils/src/lib.rs:141]:
```suggestion
                _ => panic!("Label variants cannot contain data."),
```

File: crates/bevy_macro_utils/src/lib.rs
Comment (alice-i-cecile) [inline crates/bevy_macro_utils/src/lib.rs:149]:
```suggestion
        syn::Data::Union(_) => panic!("Unions cannot be used as labels."),
```

File: crates/bevy_utils/src/label.rs
Comment (alice-i-cecile) [inline crates/bevy_utils/src/label.rs:56]: Can we collapse this back down to a single argument? I can't imagine this will ever be called in a case where you care about breaking convention on the `*Id`.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Looking good! To do:

1. Address small nits.
2. Collapse the `define_label!` macro back down to a single argument.
3. Add and run some quick schedule construction benchmarks (another PR is probably ideal).

File: crates/bevy_ecs/src/schedule/state.rs
Comment (JoJoJet) [reply crates/bevy_ecs/src/schedule/state.rs:111]: This call just adds a label of type `StateRunCriteriaLabel`, which I removed because it requires runtime labels, is private, and isn't used anywhere. All of the checks passed after removing it, so I'm pretty sure this is just dead code that rustc missed.

**Comment by JoJoJet** [comment]
> 3. Add and run some quick schedule construction benchmarks (another PR is probably ideal).

Do you mean another PR in addition to this one or to replace this one?

**Comment by alice-i-cecile** [comment]
As commented on Discord; I meant another PR.

**Review by alice-i-cecile — APPROVED** [review]
Approved pending not-terrible benchmark results :) I think this is a clearer and simpler way to get the same benefits.

**Comment by JoJoJet** [comment]
#4961

**Comment by JoJoJet** [comment]
The benchmarks are very promising (located [here](https://github.com/JoJoJet/bevy/blob/bfb7d2fadaf6e1ca24fb99f43cbb9bfcb60458d5/benches/benches/bevy_ecs/schedule.rs))

## Old labels
| # of Systems | No Deps | W/ Deps | 
|-|-|-|
| 100 | 747.62 us | 3.4763 ms 
| 500 | 1.2992 ms | 88.073 ms 
| 1000 | 1.9776 ms | 420.98 ms 

## New Labels
| # of Systems| No Deps | W/ Deps |
|-|-|-|
| 100 | 741.35 us | 2.4197 ms
| 500 | 1.2930 ms | 61.667 ms
| 1000 | 1.9627 ms | 299.96 ms

## % Change
| # of Systems | No Deps | W/ Deps |
|-|-|-|
| 100 | _-0.5121%_ | -30.201%
| 500 | _-0.5987%_ | -29.982%
| 1000 | _-0.9030%_ | -28.746%

**Comment by alice-i-cecile** [comment]
Excellent! Those are the sort of numbers I like to see. This is a one-time cost at app-startup (until #279 lands), so it's not as critical as some of the other perf optimizations, but it's still really nice to shave down app initialization time when we can.

File: crates/bevy_macro_utils/src/lib.rs
Comment (cart) [inline crates/bevy_macro_utils/src/lib.rs:128]: Given that `as_str()` results are used for comparison, it seems like we could cut down on Eq + Hash costs by:
1. Using an empty string for Unit structs
2. Only using the enum variant name for Enums

For debugging purposes, the full type name could be used. It would mean storing a separate type_name string, which increases the size of each label by an extra `&'static str`. But its worth checking to see if its worth it.

Comment (JoJoJet) [reply crates/bevy_macro_utils/src/lib.rs:128]: That's definitely something to look into but I think it would make sense to discuss it as a separate PR due to all the tradeoffs you have to make with micro-optimizations like that.

Comment (JoJoJet) [reply crates/bevy_macro_utils/src/lib.rs:128]: I thought of a simple, zero-cost optimization for this. When one side of an `==` operation is a concrete `impl SystemLabel` and the other is a type erased `SystemLabelId`, we can allow the concrete type to perform a custom comparison. For unit types, this means really fast comparisons by ignoring the string. For enums, this would mean chopping off the type name part of the string and only comparing the variant name.

Granted this wouldn't be too meaningful, since most label comparisons are done between two type erased labels. Although there are some, such as when inserting a new `Stage`, I believe.

File: crates/bevy_ecs/src/schedule/state.rs
Comment (hymm) [reply crates/bevy_ecs/src/schedule/state.rs:111]: aren't these labels for if someone wants to order around these systems?

Comment (JoJoJet) [reply crates/bevy_ecs/src/schedule/state.rs:111]: > aren't these labels for if someone wants to order around these systems?

That was probably the intention, but as far as I can tell the labels were never exposed outside of the module, so they were never actually used for that purpose. The dead code lint got suppressed because the type had traits implemented for it, but in practice it was never used.

File: crates/bevy_macro_utils/src/lib.rs
Comment (JoJoJet) [reply crates/bevy_macro_utils/src/lib.rs:128]: I did end up trying this change, and the benchmarks were not promising. Best improvement was -2%, worst regression was +5%. The others weren't affected at all.

File: crates/bevy_hierarchy/src/lib.rs
Comment (alice-i-cecile) [inline crates/bevy_hierarchy/src/lib.rs:34]: Why did this get included in this PR? Does CI break without it?

Comment (JoJoJet) [reply crates/bevy_hierarchy/src/lib.rs:34]: Good question, that must have shown up while I was rebasing. Not sure why.

File: crates/bevy_ecs/src/lib.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/lib.rs:37]: I don't think we want to add the new types to the prelude? They really don't seem useful enough.

Comment (JoJoJet) [reply crates/bevy_ecs/src/lib.rs:37]: Seems reasonable, it's easier to add them later if we end up wanting to.

**Comment by alice-i-cecile** [comment]
bors retry

**Pull request closed without merging**
